File: Day01/python/main.py
def read_input_as_columns() -> (list[int], list[int]):
    with open("input.txt") as file:
        lines = file.read().rstrip().split('\n')
    def convert_line(line):
        [x, y] = line.split()
        return [int(x), int(y)]
    lines = map(convert_line, lines)
    return tuple(map(list, zip(*lines)))

def part1():
    (column0, column1) = read_input_as_columns()
    side_by_side = zip(sorted(column0), sorted(column1))
    total_distance = sum(abs(x - y) for (x, y) in side_by_side)
    print(total_distance)

def part2():
    (column0, column1) = read_input_as_columns()
    def similarity(x):
        return x*column1.count(x)
    total_similarity = sum(similarity(x) for x in column0)
    print(total_similarity)

if __name__ == '__main__':
    #part1()
    part2()

File: Day01/haskell/main.hs
import Data.List (sort)

readInputAsColumns :: IO ([Int], [Int])
readInputAsColumns = do
  contents <- readFile "../input.txt"
  let parseLine :: String -> (Int, Int)
      parseLine = (\[a, b] -> (a, b)) . map read . words
  let parsedLines = map parseLine (lines contents) :: [(Int, Int)]
  return $ unzip parsedLines

part1 :: IO ()
part1 = do
  (column0, column1) <- readInputAsColumns
  let totalDistance = sum $ zipWith (\x y -> abs (x - y)) (sort column0) (sort column1)
  print totalDistance

part2 :: IO ()
part2 = do
  (column0, column1) <- readInputAsColumns
  let count :: Eq a => a -> [a] -> Int
      count _     [] = 0
      count x (y:ys) = (if x == y then 1 else 0) + count x ys
  let totalSimilarity = sum $ map (\x -> x * count x column1) column0
  print totalSimilarity  

main :: IO ()
main = do
  part1
  part2
